Insert nvtx markers into generated triton kernels

alexmsettle opened the pull request:
Fixes #103552 Adds nvtx markers around the generated triton kernel call functions.  Provides descriptive information about the original operators that were fused by the kernel.  Useful for performance analysis and general understanding of how the fusion algorithm works.

Information about the originating operators is appended to the RecordFunctionFast profiling message in the CachingAutotuner.  The OriginOp message is a list consisting of op_type, module_name, torch_op, seq_nr.  Each operator in the list is one of the ops that was included in the fused triton kernel.  The OriginOp message is enabled by default in the same nvtx marker as the kernel_name string.  There is an additional logic path for the case when the kernel is called from the extern_kernel() function.  In this case, an nvtx_range push/pop pair is inserted around the kernel launch in the generated inductor code.  The nvtx ranges are only inserted when **torch._inductor.config.profiler_mark_wrapper_call** is enabled.

Example timeline visualizations are included below.

For the extern kernel case a new marker is added to show details about the operator such as the unique name and sequence number.  The sequence number is used to link a forward op to its corresponding backward op.


![kineto_extern_kernels](https://github.com/pytorch/pytorch/assets/37422826/fea63cc5-acd4-472c-bc8b-f320f87774b2)

This is the backward convolution which corresponds to the above forward convolution, the sequence_nr == 132 is how they are associated with each other.

![kineto_conv_backward](https://github.com/pytorch/pytorch/assets/37422826/72c60c62-169b-4b84-9746-4aed0776d39c)

Here's an example with Nsight Systems running Resnet50.  You can see the OriginOp message is appended to the triton_kernel name.  The unique layer names for the ops that contribute to the fused kernel show clearly where they are located in the network architecture.  eg **L.layer3.0.bn1 + L.layer3.0.relu**

![nsys_rn50_triton_example](https://github.com/pytorch/pytorch/assets/37422826/f33c04ee-c8a0-4415-be29-6e6d90ab3a4b)


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang @Xia-Weiwen @ngimel

alexmsettle committed:
Insert nvtx markers into generated triton kernels

alexmsettle committed:
Fixed conflicts

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/codegen/triton.py

alexmsettle committed:
Lintrunner fixes

alexmsettle committed:
CI fixes

alexmsettle committed:
Fixed merge conflict

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/scheduler.py

alexmsettle committed:
Fxied a CI error

davidberard98 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -513,8 +518,22 @@ def generate(self):
+                        if origin_info_list:
+                            marker_string = str(origin_info_list)
+                            self.wrapper_call.writeline(
+                                f'with record_function("triton_info: {marker_string}"):'
```
fyi, if this is torch.profiler.record_function - we're seeing issues with high overhead from this. It's probably fine if you gate it with a config. Or, I'm also working on https://github.com/pytorch/pytorch/pull/106351 which should have lower overhead (at least when the profiler is not on)

alexmsettle replied:
Hi @davidberard98 - I just got my code working with your changes to fast record_function.  Here's an example screen shot.  I haven't had much time to work on this lately, but I think it is getting close now.

![image](https://github.com/pytorch/pytorch/assets/37422826/0250c37c-6b4d-435e-b4c5-56acc54c8006)

alexmsettle commented:
Thanks for the heads up!


On Thu, Aug 17, 2023 at 8:01 PM David Berard ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In torch/_inductor/codegen/wrapper.py
> <https://github.com/pytorch/pytorch/pull/103644#discussion_r1297935542>:
>
> >                  else:
> +                    # Inspect the contents of line to see if
> +                    # it is a triton kernel.  If so add the record_func
> +                    with contextlib.ExitStack() as k_stack:
> +                        if origin_info_list:
> +                            marker_string = str(origin_info_list)
> +                            self.wrapper_call.writeline(
> +                                f'with record_function("triton_info: {marker_string}"):'
>
> fyi, if this is torch.profiler.record_function - we're seeing issues with
> high overhead from this. It's probably fine if you gate it with a config.
> Or, I'm also working on #106351
> <https://github.com/pytorch/pytorch/pull/106351> which should have lower
> overhead (at least when the profiler is not on)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/103644#pullrequestreview-1583757705>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI5QN2W5NKSOEHZA4PZEH43XV3LIVANCNFSM6AAAAAAZHEXTBI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Fixed new triton profile logic so it inserts

markers before each kernel call

alexmsettle committed:
Updated sequence_nr test case

alexmsettle committed:
Cleaned up nvtx marker generation code.

Fixed lint errors.

alexmsettle committed:
Added guards to only enable markres when

config option profiler_mark_wrapper_call is set

alexmsettle committed:
Fixed failing CI test cases.

alexmsettle committed:
Fixed lint error

alexmsettle committed:
Fixed conflicts from upstream merge

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	test/inductor/test_aot_inductor.py

alexmsettle committed:
Fixed failing test case

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Fixed conflicts

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/codegen/triton.py
	torch/_inductor/utils.py

alexmsettle committed:
Syncing to remote origin

Merge remote-tracking branch 'origin/alex_nvtx_triton' into alex_nvtx_triton

alexmsettle committed:
Updated the code to use the fast record_function

that was introduced in triton_heuristics.py

alexmsettle committed:
Lint fixes

alexmsettle committed:
Fixed CI error

alexmsettle committed:
Switched from record_function to nvtx_range

push / pop for adding markers to the external kernels

alexmsettle commented:
Hi @davidberard98  I think I'm ready to start the code review for this one.

alexmsettle committed:
Fixed regression error.

alexmsettle committed:
Lintrunner fixes

alexmsettle committed:
Fixed conflicts

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/codegen/triton.py
	torch/_inductor/codegen/wrapper.py
	torch/_inductor/scheduler.py

alexmsettle commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2574,11 +2589,17 @@ def define_kernel(self, src_code, node_schedule):
             # to "triton_" to maximize caching opportunities (when unique_kernel_names = False).
             src_code = src_code.replace(str(Placeholder.DESCRIPTIVE_NAME), kernel_name)
             src_code = src_code.replace(str(Placeholder.KERNEL_NAME), subs_name)
+            src_code = src_code.replace(f"'{Placeholder.ORIGIN_INFO}'", str(op_info))
```
Hi @davidberard98 I noticed while getting regressions to pass that these Placeholders are used to to prevent triton from generating multiple copies of the same kernel if the operator instance resolves to the same kernel implementation as a previous operator instance.  See this test [test_dropout3](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor.py#L5704-L5710).  It has 2 instances of a dropout op.  The test expects to see 1 dropout kernel which is called once per drop out instance.  When we put information in the meta attribute of the triton kernel, like the kernel name with an instance id, it will cause a new kernel instance to be generated.  I see the placeholders are used to prevent this from happening, but you only get 1 kernel instance recorded, so you're losing information about the 2nd kernel instance.
In my case I want to see information specific to each instance of the kernel.  The origin_op info is showing specifically which operator instances in the network graph were fused together for this specific instance of the kernel.   So I think I need to capture the op info at the kernel call, not the kernel definition.

davidberard98 replied:
hey, thanks for the info (and sorry for the delayed review!)

> but you only get 1 kernel instance recorded, so you're losing information about the 2nd kernel instance.
is this still the case when descriptive names are turned on?

if you go the route of capturing the info at the kernel call, do you have any ideas how to do that?

alexmsettle replied:
Oh ok let me try with descriptive names turned on.
So far the only way I know to capture info at the time of kernel call is to insert an nvtx_range push/pop.  
I'll get back to you soon, I'm kind of swamped at work right now so my responses will be delayed.

alexmsettle replied:
Changing the value of descriptive_names and unique_kernel_names didn't have any effect on the number of instances of a kernel.  I'd like to just go with the current approach and live with inserting the marker at the kernel definition rather than the kernel call.

davidberard98 commented on torch/_inductor/utils.py:
```diff
@@ -357,34 +360,79 @@ def aggregate_origins(node_schedule):
 
 
-def get_fused_kernel_name(node_schedule, descriptive_names):
+def get_module_name_from_meta(meta_dict):
```
reminder to myself, will take a closer look at this logic later

alexmsettle replied:
Done - I added several comments and made the regex simpler.

eg.   
       Before: 'getattr_L__self___layer4___0___downsample_1'
        After: L.layer4.0.downsample_1

alexmsettle replied:
I just added a test case that checks the module name after the name filtering is applied.
**test_inductor_extern_kernel_mod_name()**

alexmsettle committed:
Fixed merge conflict

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/utils.py

alexmsettle committed:
Cleaned up dead code from earlier revisions

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Added test cases to protect new nvtx markers

alexmsettle commented:
Hi @davidberard98 this is ready for review.  I have been really busy at work and haven't had much time to focus on this PR.  Thanks for your patience.

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

davidberard98 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -170,6 +170,8 @@ def __init__(
         self.record_function_ctx = torch._C._profiler._RecordFunctionFast(
             self.meta.get("kernel_name", "triton kernel")
+            + ", OriginOps: "
+            + str(self.meta.get("origin_ops", []))
```
ideally we want these to go in the metadata instead of appearing directly in the name of the event... but for now this is probably fine

alexmsettle replied:
I was looking into how you could make the origin_ops or any other attribute go into the nvtx markers in a field other than name.  The nvtx range API provides a 'payload' field that you can set with string or numeric values.  The problem is that you need to use a different function than is currently used in pytorch.  The function we want is [nvtxRangePushEx](https://docs.nvidia.com/nsight-visual-studio-edition/nvtx/index.html). The user can pass in an eventAttrib where you can set the payload.  Currently pytorch uses the[ nvtxRangePushA()](https://github.com/pytorch/pytorch/blob/main/torch/csrc/profiler/stubs/cuda.cpp#L78) function which only accepts a string message as an argument.
For Kineto, I think the [RecordFunction::before() ](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/record_function.h#L289-L309)method would need to be modified in order to support an extra string type argument to carry the origin_ops or other optional attributes.

davidberard98 commented on test/inductor/test_profiler.py:
```diff
@@ -75,6 +75,7 @@ def _test_profiling_kernel_names(self, fn, args, kernel_name_str: str):
                     hasattr(event, "name")
                     and kernel_name_str in event.name
                     and "triton" in event.name
+                    and "OriginOps" in event.name
```
Can you add a test that does `export_chrome_trace` on a trace that has OriginOps in it? Reason: the json serialization code is a bit buggy, so it would be good to have a sanity check that the OriginOps format doesn't expose any serialization bugs

alexmsettle replied:
done.
Added **test_profiling_origin_ops_with_chrome_json**

davidberard98 reviewed (commented):
Generally LGTM (a few nits in comments inline) - but I had a high-level question:

* It seems like extern_kernels get wrapped with nvtx markers at the callsite instead of the definition
* All other kernels either don't have a nvtx marker, or it appears just as part of the event name

is that correct?

also, would be great if you can add some screenshots in the PR to demonstrate what it looks like, so we can point people to the PR if for instructions on how to use it

alexmsettle commented:
Thanks for reviewing @davidberard98 , I'm working on some updates.

alexmsettle committed:
Added a new test case for chrome trace

Cleaned up text in the nvtx markers

alexmsettle committed:
Fixed failing test cases

alexmsettle committed:
Code review cleanup.

Simplified the regex for parsing the unique layer names and
pytorch op types.

alexmsettle commented:
> Generally LGTM (a few nits in comments inline) - but I had a high-level question:
> 
> * It seems like extern_kernels get wrapped with nvtx markers at the callsite instead of the definition
> * All other kernels either don't have a nvtx marker, or it appears just as part of the event name
> 
> is that correct?
> 
> also, would be great if you can add some screenshots in the PR to demonstrate what it looks like, so we can point people to the PR if for instructions on how to use it

Yes extern_kernels and fallback_kernels get wrapped with record_function() at the call site.  I couldn't figure out how to add the marker at the definition.  I don't see any generated code for the definition of these kernels.

Yes all the triton kernels have marker information appended to the kernel name string, formatted like this.
**OriginOps: [{'op_type': 'addmm', 'mod_name': 'L.linear_list_1', 'torch_op': 'torch.nn.modules.linear.Linear', 'seq_nr': 151}]**

Any remaining kernels don't have a marker.

I added a few screen shots from both kineto and nsys to the PR description.

alexmsettle commented on torch/_inductor/select_algorithm.py:
```diff
@@ -117,6 +117,7 @@ def jit_line(self):
             "device": V.graph.scheduler.current_device.index,
             "device_type": V.graph.scheduler.current_device.type,
             "constants": {},
+            "origin_ops": {},
```
Not sure if I should just remove it or add in TODO like you suggested.  I don't know what a **template** or **foreach** triton kernel look like.  Or more importantly, do you have an example of network code that would cause these kinds of kernels to be generated?

alexmsettle replied:
I just went ahead and removed it.

alexmsettle committed:
Added test case to check the kernel mod_name

string format is correct.
Removed unused code from select_algorithm

alexmsettle commented:
Hi @davidberard98 - I think I covered all the points you raised in the review.  It should be getting close now.

alexmsettle committed:
Fixed merge conflict from main

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/scheduler.py

alexmsettle committed:
Commit that caused original merge conflict

was reverted after I fixed the conflict.  Since the
change was reverted it caused a new conflict with my
changes that resolved the original conflict.

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/scheduler.py

alexmsettle commented:
Finally got all the tests to pass.

There is a change set (#108193) in **torch/_inductor/scheduler.py** that is conflicting with my changes.  It was committed, then reverted.  So when it re-lands there will be another conflict with my changes.

alexmsettle commented:
ping @davidberard98 can you pls review my latest changes?

davidberard98 commented:
@alexmsettle thanks for the changes - overall looks pretty good to me. a few comments:

* Another high-level question; does nvtx just get the 
* I'm going to check with some other inductor experts for opinions, but we might want to guard this behind a different inductor config - will get back to you about this (particularly for the extern kernels, where the record_function is probably adding non-negligible overhead).
* might also be worth checking if you can replace record_function with RecordFunctionFast; it should reduce the overhead a bit

also follow-up to your question from last week that I missed:
> I don't know what a template or foreach triton kernel look like. Or more importantly, do you have an example of network code that would cause these kinds of kernels to be generated?

template: used for matmuls and other similar kernels (they're called "templates" because, unlike the pointwise/reduction kernels that are entirely code-generated by inductor, matmuls require the developer to write a template that implements the matmul and provides stubs at the end where an epilogue can be filled in)
foreach: used for "foreach kernels" like foreach_add, which run a given operation (e.g. add) over a list of elements; used in optimizers (e.g. SGD) where there's a lot of different tensors that need the same operation applied over all the tensors.

examples in these profiler tests:
https://github.com/pytorch/pytorch/blob/a1e222ef02f07d8c2626d635d6c6b87401816d92/test/inductor/test_profiler.py#L93-L119

alexmsettle committed:
Fixed conflicts

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/scheduler.py

alexmsettle committed:
Switched to use RecordFunctionFast in generated

inductor code.

alexmsettle committed:
Fixed failing CI test

alexmsettle committed:
Fixed conflicts

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/codegen/triton.py
	torch/_inductor/triton_heuristics.py

alexmsettle commented:
> @alexmsettle thanks for the changes - overall looks pretty good to me. a few comments:
> 
> * Another high-level question; does nvtx just get the
> * I'm going to check with some other inductor experts for opinions, but we might want to guard this behind a different inductor config - will get back to you about this (particularly for the extern kernels, where the record_function is probably adding non-negligible overhead).
> * might also be worth checking if you can replace record_function with RecordFunctionFast; it should reduce the overhead a bit
> 
> also follow-up to your question from last week that I missed:
> 
> > I don't know what a template or foreach triton kernel look like. Or more importantly, do you have an example of network code that would cause these kinds of kernels to be generated?
> 
> template: used for matmuls and other similar kernels (they're called "templates" because, unlike the pointwise/reduction kernels that are entirely code-generated by inductor, matmuls require the developer to write a template that implements the matmul and provides stubs at the end where an epilogue can be filled in) foreach: used for "foreach kernels" like foreach_add, which run a given operation (e.g. add) over a list of elements; used in optimizers (e.g. SGD) where there's a lot of different tensors that need the same operation applied over all the tensors.
> 
> examples in these profiler tests:
> 
> https://github.com/pytorch/pytorch/blob/a1e222ef02f07d8c2626d635d6c6b87401816d92/test/inductor/test_profiler.py#L93-L119

Thanks for reviewing @davidberard98 

- I replaced record_function() with RecordFunctionFast()
- Your question about nvtx was cut off, I didn't understand.   I found that nvtx markers are directly tied to nsight systems, so they won't show up in the kineto trace.  That's why I went back to RecordFunction, I wanted this to work for either profiler. RecordFunction automatically inserts nvtx markers at some point, but that happens after the kineto trace is captured.
- I'll review the triton template and foreach, if it's straight forward I can go ahead and add the markers.  Looks like this path would follow the same path as the triton_kernels.

alexmsettle committed:
Added origin op info to triton template kernels

and foreach kernels

alexmsettle commented:
FYI - I just added support for the template and foreach kernels.
![image](https://github.com/pytorch/pytorch/assets/37422826/e32dd238-60a8-4774-83ff-75e936bab844)

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle commented:
Hi @davidberard98 I think this PR is just waiting to see which inductor config option I should use to control when the external_kernel markers are inserted using RecordFunctionFast.  If there are any other changes you'd like to see please let me know.  Thanks!

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle commented:
hi @davidberard98 - are there any other changes you need me to make for this?  We're getting close to the holidays.  Hoping to wrap it up before everyone takes off.  I think you wanted to confirm what config option I should use to guard the extern_kernel markers.

davidberard98 commented:
@alexmsettle I will chat with people tomorrow

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -120,7 +120,10 @@ def jit_line(self):
-        inductor_meta = {"kernel_name": str(Placeholder.DESCRIPTIVE_NAME)}
+        inductor_meta = {
+            "kernel_name": str(Placeholder.DESCRIPTIVE_NAME),
+            "origin_ops": str(Placeholder.ORIGIN_INFO),
```
Can we include `origin_ops` in the `inductor_meta` only if `profiler_mark_wrapper_call` is enabled ?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -429,35 +431,103 @@ def aggregate_origins(node_schedule):
+    return mod_name, torch_op
+
+
+def get_origin_op_info(node_schedule, descriptive_names):
```
can you add a docstring to all new functions ?

alexmsettle replied:
thanks for reviewing @eellison , working on the revisions.

eellison commented on torch/_inductor/utils.py:
```diff
@@ -429,35 +431,103 @@ def aggregate_origins(node_schedule):
+        # <class > tag.
+        # Example <class 'torch.nn.modules.conv.Conv2d'> is reduced to
+        # torch.nn.modules.conv.Conv2d
+        res = re.search(r"<class\s+'(\S+)'>", str(torch_op))
```
Can we compile / lru cache all of the regexes ?

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -592,6 +594,23 @@ def generate(self, is_inference):
+                elif (
+                    origin_info_list
+                    and isinstance(line, str)
+                    and not re.search(r"^#", line)
```
It feels off to me that we should have to do a regex search in the middle of our code generation loop. Can we figure out a better way ?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -429,35 +431,103 @@ def aggregate_origins(node_schedule):
+        #                 <class 'torch.nn.modules.batchnorm.BatchNorm2d'>)}
+        # The key is the unique module name.  The value contains the
+        # torch_op type eg. <class 'torch.nn.modules.batchnorm.BatchNorm2d'>
+        mod_name, extra = list(meta_dict["nn_module_stack"].items())[-1]
```
cc @SherlockNoMad @ydwu4 do you know if there are any nn_module_stack utilities that would replace this ?

eellison commented on torch/_inductor/utils.py:
```diff
@@ -888,6 +958,36 @@ def __exit__(self, *args):
+        marker_string = str(self)
+        prefix = f"{self.kernel_prefix}_kernel, OriginOps: "
+        code.writeline(
+            f'with torch._C._profiler._RecordFunctionFast("{prefix}{marker_string}"):'
```
@davidberard98 assuming youve looked at this

eellison reviewed (commented):
looks good, have some comments

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Merge branch 'pytorch:main' into alex_nvtx_triton

alexmsettle committed:
Code review changes

Guarded updates to meta origin_ops under profiler_mark_wrapper_call

alexmsettle committed:
Resolved conflicts

Merge remote-tracking branch 'origin/master' into alex_nvtx_triton

Conflicts:
	torch/_inductor/utils.py

alexmsettle committed:
Local branch got out of sync with remote

Merge remote-tracking branch 'origin/alex_nvtx_triton' into alex_nvtx_triton

alexmsettle committed:
Fixed runtime error, python multi-line if clause

alexmsettle committed:
Lint fixes

alexmsettle committed:
Fixed test case

OriginOps is only included in the nvtx marker name when the
profiler_mark_wrapper_call config option is set.
